packages/edit-widgets/src/components/layout/unsaved-changes-warning.js

```diff
@@ -0,0 +1,52 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { useEffect } from '@wordpress/element';
+import { useSelect } from '@wordpress/data';
+
+/**
+ * Warns the user if there are unsaved changes before leaving the editor.
+ * Compatible with Post Editor and Site Editor.
+ *
+ * @return {WPComponent} The component.
+ */
+export default function UnsavedChangesWarning() {
+	const isDirty = useSelect( ( select ) => {
+		return () => {
+			const { __experimentalGetDirtyEntityRecords } = select( 'core' );
+			const dirtyEntityRecords = __experimentalGetDirtyEntityRecords();
+			return dirtyEntityRecords.length > 0;
+		};
+	}, [] );
+
+	/**
+	 * Warns the user if there are unsaved changes before leaving the editor.
+	 *
+	 * @param {Event} event `beforeunload` event.
+	 *
+	 * @return {?string} Warning prompt message, if unsaved changes exist.
+	 */
+	const warnIfUnsavedChanges = ( event ) => {
```

Comment:
Prefer to move this function to `useEffect` to make the linter happy (#24914) and avoid creating this function on every render.

Reply:
Done in 8808ca6